src/wp-includes/update.php

```diff
@@ -956,6 +956,52 @@ function wp_clean_update_cache() {
 	delete_site_transient( 'update_core' );
 }
 
+/**
+ * Deletes all contents of the temp-backup directory.
+ *
+ * @since 6.0.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ */
+function wp_delete_all_temp_backups() {
+	/*
+	 * Check if there's a lock, or if currently performing an Ajax request,
+	 * in which case there's a chance we're doing an update.
+	 * Reschedule for an hour from now and exit early.
+	 */
+	if ( get_option( 'core_updater.lock' ) || get_option( 'auto_updater.lock' ) || wp_doing_ajax() ) {
+		wp_schedule_single_event( time() + HOUR_IN_SECONDS, 'wp_delete_temp_updater_backups' );
+		return;
+	}
+
+	add_action(
+		'shutdown',
+		/*
+		 * This action runs on shutdown to make sure there's no plugin updates currently running.
+		 * Using a closure in this case is OK since the action can be removed by removing the parent hook.
+		 * `remove_action( 'wp_delete_temp_updater_backups', 'wp_delete_all_temp_backups' );`
+		 */
+		function() {
+			global $wp_filesystem;
+
+			if ( ! $wp_filesystem ) {
+				include_once ABSPATH . '/wp-admin/includes/file.php';
+				WP_Filesystem();
+			}
+
+			$dirlist = $wp_filesystem->dirlist( $wp_filesystem->wp_content_dir() . 'upgrade/temp-backup/' );
+
+			foreach ( array_keys( $dirlist ) as $dir ) {
```

Comment:
`WP_Filesystem_*::dirlist()` can return `false` should it be unable to list directory contents. While checks have already been done for prior operations, I wonder if this might still benefit from a guard against a strict `false` value on `$dirlist`. Thoughts?

Reply:
Should we simply cast $dirlist as an array?

Reply:
This done in 26dc49e30f5addfd8b085b54973ea230c1161ad0

Reply:
Protects against an error on `false` and any performance impact vs a 3-line guard minimal. 👍

Reply:
I don't think casting `$wp_filesystem->dirlist()` to an array is appropriate here.

If `dirlist()` does return false, then it's the equiv of `$dirlist = array( false )`, which will result in `$wp_filesystem->delete( $temp_backup_dir . false, true );`, which will delete the entire `temp-backup`.

Reply:
could do a 1-line guard as:

```
$dirlist         = (array) $wp_filesystem->dirlist( $temp_backup_dir );
$dirlist         = $dirlist ? $dirlist : array();
```

Reply:
Good idea :+1:, except don't cast the first line as an array. You'll get `array( false )`, which is truthy.

I believe the `|| 0 === $dir` can then be removed too.

Reply:
yea, I meant to remove that cast :-)

Reply:
> I don't think casting `$wp_filesystem->dirlist()` to an array is appropriate here.
> 
> If `dirlist()` does return false, then it's the equiv of `$dirlist = array( false )`, which will result in `$wp_filesystem->delete( $temp_backup_dir . false, true );`, which will delete the entire `temp-backup`.
> 

@pbiron FYI, the `|| 0 === $dir` guard in the `foreach` protected against the issue above.

```php
$dirlist = array( false );

// array_keys( $dirlist ) will be array( 0 ).

foreach( array_keys( $dirlist ) as $dir ) {
    if ( 0 === $dir ) {
        continue; // successfully skipped.
    }
}
```

Reply:
there are a number of other cases within core where the return value of `$wp_filesystem->dirlist()` is iterated over without checking whether it returned false (many of them in the upgrader).  I'll open another ticket to address those cases (which we should **not** fix as part of this PR)

Reply:
So removing the cast, add the guard, remove the check inside the foreach.

Reply:
updated b6b9d77e793a96540cb77eb13d8cd31bad39d7e0